File: typescript/square.ts
class SquareRoot {
    private num: number;
    private precision: number;
    
    constructor(N: number, precision?: number) {
        precision = precision || 0.00000001;

        this.num = N;
        this.precision = precision;
    }

    public CalculateEstimation():number {
        let countDigits:number = 0;
        let copyForCountDigits = this.num;

        let divisionDigits:number[] = [];

        while (copyForCountDigits > 0){
            divisionDigits.unshift(copyForCountDigits);
            copyForCountDigits /= 10;
            countDigits++;
        }

        if (divisionDigits.length == 0) return 0;

        let rootBase:number = divisionDigits[0];
        let exponent:number = countDigits - 1;

        if (countDigits % 2 == 0){
            rootBase = divisionDigits[1];
            exponent = countDigits - 2;
        }

        return rootBase < 10 ? 2 * Math.pow(10, exponent / 2) : 6 * Math.pow(10, exponent / 2);
    }

    public sqrt():number {
        if (this.num == 0) return 0;

        let currentPrecision = 1;
        let estimation:number = this.CalculateEstimation();

        while(currentPrecision > this.precision){
            estimation = (this.num/estimation + estimation)/2;
            currentPrecision = Math.abs(estimation * estimation - this.num);
        }

        return estimation;
    }
}

let N = 5000000;

let results:number[] = [];

let start = new Date();

for(let i=0; i<N;i++){
    results.push(new SquareRoot(i).sqrt());
}

let end = new Date();

console.log("Results[4250250] = " + results[4250250]);
console.log("Total elapsed " + (end.getTime() - start.getTime()) + " ms");

/*
results.forEach(function(val: number, ind: number){
    console.log(ind + ". " + val);
})*/

File: kotlin/src/Square.kt
import java.util.ArrayList

class SquareRoot constructor(private val Num: Int, private val Precision: Double = 0.00000001) {

    private fun calculateEstimation(): Double {
        var copyForCountDigits = Num
        var countDigits = 0

        val divisionDigits = ArrayList<Int>()

        while (copyForCountDigits > 0) {
            divisionDigits.add(0, copyForCountDigits)

            copyForCountDigits /= 10
            countDigits++
        }

        if (divisionDigits.size == 0) return 0.0

        var rootBase = divisionDigits[0]
        var exponent = countDigits - 1

        if (countDigits % 2 == 0) {
            rootBase = divisionDigits[1]
            exponent = countDigits - 2
        }

        return if (rootBase < 10) 2 * Math.pow(10.0, (exponent / 2).toDouble()) else 6 * Math.pow(
            10.0,
            (exponent / 2).toDouble()
        )
    }

    fun sqrt(): Double {
        if (Num == 0) return 0.0

        var currentPrecision = 1.0
        var estimation = calculateEstimation()

        while (currentPrecision > Precision) {
            estimation = (Num / estimation + estimation) / 2
            currentPrecision = Math.abs(estimation * estimation - Num)
        }

        return estimation
    }
}

fun main(args: Array<String>) {
    val N = 5000000
    val results = DoubleArray(N)

    val start = System.nanoTime()
    for (i in 0 until N) {
        results[i] = SquareRoot(i).sqrt()
    }
    val finish = System.nanoTime()

    println("Results[4250250] = " + results[4250250])
    println("Time elapsed: " + (finish - start) / 1000000000.0)
}


